src/relay/dnsrelay/mod.rs

```diff
@@ -225,8 +143,22 @@ pub async fn run(context: SharedContext) -> io::Result<()> {
         }
     });
 
+    let config = context.config();
+    #[cfg(target_os = "android")]
+    let local_upstream = Arc::new(upstream::UnixSocketUpstream {
+        path: config.local_dns_path.clone().expect("local query DNS path"),
+    });
+    #[cfg(not(target_os = "android"))]
+    let local_upstream = Arc::new(upstream::UdpUpstream {
+        server: config.local_dns_addr.clone().expect("local query DNS address"),
+    });
     // FIXME: We use TCP to send remote queries by default, which should be configuable.
     let balancer = PlainPingBalancer::new(context.clone(), ServerType::Tcp).await;
```

Comment:
@zonyitoo Shouldn't we use `Arc::clone` for `SharedContext`?

Reply:
What do you mean? `context.clone()` calls `SharedContext::clone`, which is `Arc::clone`. No?

Reply:
I see you are right. https://doc.rust-lang.org/std/sync/struct.Arc.html#cloning-references